Bug: for_each on non-iterable types passes type checking

## Description
The type checker doesn't validate that `for_each` is used with an iterable type (array or string). Non-iterable types like maps, ints, and structs pass type checking but fail at runtime with E3017.

## Reproduction
```ez
import @std
using std

do main() {
    temp m map[string:int] = {"a": 1}
    for_each item in m {  // Should be compile error, but passes
        println(item)
    }
}
```

Also fails for:
- `for_each` on `int`
- `for_each` on `struct`

All pass `ez check` but fail at runtime with:
```
error[E3017]: for_each requires array or string, got MAP
```

## Expected Behavior
The type checker should validate that `for_each` is only used with arrays or strings and produce a compile-time error for other types.

## Root Cause
The type checker is not validating the type of the iterable expression in `for_each` statements.

--- a/integration-tests/fail/errors/E3017_for_each_non_iterable.ez
+++ b/integration-tests/fail/errors/E3017_for_each_non_iterable.ez
@@ -1,0 +1,14 @@
+/*
+ * Error Test: E3017 - for_each-non-iterable
+ * Expected: "for_each requires" or "array or string"
+ */
+
+import @std
+using std
+
+do main() {
+    temp m map[string:int] = {"a": 1}
+    for_each item in m {  // Should fail: map is not iterable
+        println(item)
+    }
+}

--- a/pkg/typechecker/typechecker.go
+++ b/pkg/typechecker/typechecker.go
@@ -2411,17 +2411,26 @@ func (tc *TypeChecker) checkForStatement(forStmt *ast.ForStatement, expectedRetu
 func (tc *TypeChecker) checkForEachStatement(forEach *ast.ForEachStatement, expectedReturnTypes []string) {
 	tc.enterScope()
 
-	// Infer element type from collection
+	// Infer element type from collection and validate it's iterable (#595)
 	if forEach.Variable != nil && forEach.Collection != nil {
 		collType, ok := tc.inferExpressionType(forEach.Collection)
 		if ok {
 			// For arrays, element type is inside []
-			if len(collType) > 2 && collType[0] == '[' {
-				elemType := collType[1 : len(collType)-1]
+			if tc.isArrayType(collType) {
+				elemType := tc.extractArrayElementType(collType)
 				tc.defineVariable(forEach.Variable.Value, elemType)
 			} else if collType == "string" {
 				// Iterating over string gives char
 				tc.defineVariable(forEach.Variable.Value, "char")
+			} else {
+				// Not an iterable type - produce error
+				line, column := tc.getExpressionPosition(forEach.Collection)
+				tc.addError(
+					errors.E3017,
+					fmt.Sprintf("for_each requires array or string, got %s", collType),
+					line,
+					column,
+				)
 			}
 		}
 	}

--- a/pkg/typechecker/typechecker_test.go
+++ b/pkg/typechecker/typechecker_test.go
@@ -1365,7 +1365,7 @@ do main() {
 	assertNoErrors(t, tc)
 }
 
-func TestForEachStatementMap(t *testing.T) {
+func TestForEachStatementMapError(t *testing.T) {
 	input := `
 do main() {
 	temp m map[string:int] = {"a": 1, "b": 2}
@@ -1374,7 +1374,7 @@ do main() {
 }
 `
 	tc := typecheck(t, input)
-	assertNoErrors(t, tc)
+	assertHasError(t, tc, errors.E3017) // for_each on map should be an error (#595)
 }
 
 // ============================================================================

